File: agent/motionDetect.ts
import type { AgentConfig } from "./index";

/**
 * ONVIF-based motion detection for cameras.
 *
 * Connects to each camera's ONVIF service, subscribes to motion alarm events,
 * and POSTs detected motion to the cloud API. Cameras without ONVIF credentials
 * or that don't support ONVIF events are silently skipped — RTSP streaming
 * continues to work regardless.
 */

// Motion-related ONVIF topics we listen for
const MOTION_TOPICS = [
  "RuleEngine/CellMotionDetector/Motion",
  "RuleEngine/MotionRegionDetector/Motion",
  "VideoAnalytics/Motion",
  "VideoSource/MotionAlarm",
  "Device/Trigger/DigitalInput",
];

// Debounce window per camera (ms)
const DEBOUNCE_MS = 10_000;

// Rate-limit tracking: cameraId -> list of timestamps
const lastMotionTimestamps: Map<string, number> = new Map();

// Track active ONVIF camera connections for cleanup
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const activeCams: Array<any> = [];
let stopped = false;

/**
 * Wrap the callback-based `Cam` constructor in a Promise.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function connectCamera(opts: { hostname: string; port: number; username: string; password: string }): Promise<any> {
  // Dynamic import so the module is optional at runtime
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const { Cam } = require("onvif");

  return new Promise((resolve, reject) => {
    const cam = new Cam(
      {
        hostname: opts.hostname,
        port: opts.port,
        username: opts.username,
        password: opts.password,
      },
      (err: Error | null) => {
        if (err) return reject(err);
        resolve(cam);
      }
    );
  });
}

/**
 * Extract the hostname (IP) from an RTSP URL.
 * e.g. rtsp://192.168.1.100:554/stream1 → 192.168.1.100
 */
function extractHostFromRtsp(rtspUrl: string): string | null {
  try {
    // Replace rtsp:// with http:// so URL parser works
    const url = new URL(rtspUrl.replace(/^rtsp:\/\//, "http://"));
    return url.hostname || null;
  } catch {
    return null;
  }
}

/**
 * Check if an ONVIF event topic is motion-related.
 */
function isMotionTopic(topicString: string): boolean {
  if (!topicString) return false;
  return MOTION_TOPICS.some((t) => topicString.includes(t));
}

/**
 * POST a motion event to the cloud API.
 */
async function reportMotion(
  config: AgentConfig,
  cameraId: string
): Promise<void> {
  const url = `${config.apiUrl}/api/motion`;
  const body = JSON.stringify({
    cameraId,
    schoolId: config.schoolId,
    timestamp: new Date().toISOString(),
  });

  try {
    const res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-School-ID": config.schoolId,
        "X-API-Key": config.apiKey,
      },
      body,
    });

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      console.warn(`[Motion] Cloud API responded ${res.status}: ${text}`);
    }
  } catch (err) {
    console.warn(`[Motion] Failed to report motion for ${cameraId}:`, err);
  }
}

/**
 * Debounce check — returns true if enough time has passed since the last
 * motion event for this camera.
 */
function shouldReport(cameraId: string): boolean {
  const now = Date.now();
  const last = lastMotionTimestamps.get(cameraId) ?? 0;
  if (now - last < DEBOUNCE_MS) return false;
  lastMotionTimestamps.set(cameraId, now);
  return true;
}

/**
 * Start ONVIF motion detection for all configured cameras.
 */
export async function startMotionDetect(config: AgentConfig): Promise<void> {
  stopped = false;
  console.log("[Motion] Motion detection listener started");
  console.log(`[Motion] Monitoring ${config.cameras.length} cameras`);

  for (const camera of config.cameras) {
    // Skip cameras without ONVIF credentials
    if (!camera.onvifUser || !camera.onvifPassword) {
      console.log(
        `[Motion] ${camera.cameraId}: no ONVIF credentials, skipping ONVIF subscription`
      );
      continue;
    }

    const hostname = extractHostFromRtsp(camera.rtspUrl);
    if (!hostname) {
      console.warn(
        `[Motion] ${camera.cameraId}: could not parse hostname from RTSP URL, skipping`
      );
      continue;
    }

    // Default ONVIF port is 80
    const onvifPort = 80;

    try {
      console.log(
        `[Motion] ${camera.cameraId}: connecting to ONVIF at ${hostname}:${onvifPort}...`
      );

      const cam = await connectCamera({
        hostname,
        port: onvifPort,
        username: camera.onvifUser,
        password: camera.onvifPassword,
      });

      if (stopped) return;

      activeCams.push(cam);

      console.log(
        `[Motion] ${camera.cameraId}: ONVIF connected, subscribing to events`
      );

      // Subscribe to events
      cam.on("event", (event: Record<string, unknown>) => {
        if (stopped) return;

        try {
          // Extract topic string — the ONVIF library puts it in various places
          // depending on the version and camera
          const topicValue =
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (event as any)?.topic?._ ||
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (event as any)?.topic ||
            "";

          const topicStr = typeof topicValue === "string" ? topicValue : String(topicValue);

          if (!isMotionTopic(topicStr)) return;

          // Some cameras send IsMotion = true/false — check for false to avoid
          // reporting "motion stopped" events
          try {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const simpleItem = (event as any)?.message?.message?.data?.simpleItem;
            if (simpleItem) {
              const items = Array.isArray(simpleItem) ? simpleItem : [simpleItem];
              for (const item of items) {
                const val = item?.$?.Value ?? item?.value ?? item?.Value;
                if (
                  val === false ||
                  val === "false" ||
                  val === "0" ||
                  val === 0
                ) {
                  // Motion ended — skip
                  return;
                }
              }
            }
          } catch {
            // If we can't parse simpleItem, still report the event
          }

          // Debounce
          if (!shouldReport(camera.cameraId)) return;

          console.log(`[Motion] ${camera.cameraId}: motion detected (${topicStr})`);
          reportMotion(config, camera.cameraId);
        } catch (err) {
          console.warn(
            `[Motion] ${camera.cameraId}: error processing event:`,
            err
          );
        }
      });
    } catch (err) {
      // Camera doesn't support ONVIF or is unreachable — not fatal
      console.warn(
        `[Motion] ${camera.cameraId}: ONVIF connection failed (streaming still works):`,
        err instanceof Error ? err.message : err
      );
    }
  }
}

/**
 * Stop all ONVIF event subscriptions and clean up.
 */
export function stopMotionDetect(): void {
  stopped = true;
  console.log("[Motion] Stopping motion detection...");

  for (const cam of activeCams) {
    try {
      // The onvif library cam objects are EventEmitters — remove listeners
      cam.removeAllListeners("event");
    } catch {
      // Ignore cleanup errors
    }
  }

  activeCams.length = 0;
  lastMotionTimestamps.clear();
  console.log("[Motion] Motion detection stopped");
}

File: agent-go/motiondetect.go
package main

import (
	"bytes"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"safeguard-agent/onvif"
)

// ONVIF motion-related topics we listen for.
var motionTopics = []string{
	"RuleEngine/CellMotionDetector/Motion",
	"RuleEngine/MotionRegionDetector/Motion",
	"VideoAnalytics/Motion",
	"VideoSource/MotionAlarm",
	"Device/Trigger/DigitalInput",
}

const motionDebounceMS = 10_000

var (
	lastMotionTimestamps   = make(map[string]time.Time)
	lastMotionTimestampsMu sync.Mutex
	motionStopped          bool
	motionStoppedMu        sync.RWMutex
	activeSubscriptions    []*onvif.EventSubscription
	activeSubscriptionsMu  sync.Mutex
)

// isMotionTopic checks if an ONVIF event topic is motion-related.
func isMotionTopic(topic string) bool {
	if topic == "" {
		return false
	}
	for _, t := range motionTopics {
		if strings.Contains(topic, t) {
			return true
		}
	}
	return false
}

// shouldReportMotion debounce check — returns true if enough time has passed.
func shouldReportMotion(cameraID string) bool {
	lastMotionTimestampsMu.Lock()
	defer lastMotionTimestampsMu.Unlock()

	now := time.Now()
	if last, ok := lastMotionTimestamps[cameraID]; ok {
		if now.Sub(last) < time.Duration(motionDebounceMS)*time.Millisecond {
			return false
		}
	}
	lastMotionTimestamps[cameraID] = now
	return true
}

// reportMotion POSTs a motion event to the cloud API.
func reportMotion(cfg *AgentConfig, cameraID string) {
	payload := map[string]string{
		"cameraId":  cameraID,
		"schoolId":  cfg.SchoolID,
		"timestamp": time.Now().UTC().Format(time.RFC3339),
	}

	body, err := json.Marshal(payload)
	if err != nil {
		fmt.Printf("[Motion] Failed to marshal motion event: %s\n", err)
		return
	}

	reqURL := cfg.APIUrl + "/api/v1/motion"
	req, err := http.NewRequest("POST", reqURL, bytes.NewReader(body))
	if err != nil {
		fmt.Printf("[Motion] Failed to create request: %s\n", err)
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-School-ID", cfg.SchoolID)
	req.Header.Set("X-API-Key", cfg.APIKey)

	resp, err := httpClient.Do(req)
	if err != nil {
		fmt.Printf("[Motion] Failed to report motion for %s: %s\n", cameraID, err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		fmt.Printf("[Motion] Cloud API responded %d: %s\n", resp.StatusCode, string(respBody))
	}
}

// extractHostFromRtsp extracts the hostname from an RTSP URL.
func extractHostFromRtsp(rtspURL string) string {
	// Replace rtsp:// with http:// so URL parser works
	httpURL := strings.Replace(rtspURL, "rtsp://", "http://", 1)
	parsed, err := url.Parse(httpURL)
	if err != nil {
		return ""
	}
	return parsed.Hostname()
}

// handleONVIFEvent processes a raw ONVIF notification message for a camera.
func handleONVIFEvent(cfg *AgentConfig, cameraID string, data []byte) {
	motionStoppedMu.RLock()
	if motionStopped {
		motionStoppedMu.RUnlock()
		return
	}
	motionStoppedMu.RUnlock()

	// Parse the ONVIF notification envelope
	var env onvif.NotificationEnvelope
	if err := xml.Unmarshal(data, &env); err != nil {
		// Try a simpler string-based approach
		content := string(data)
		for _, topic := range motionTopics {
			if strings.Contains(content, topic) {
				// Check for false/0 values indicating motion stopped
				if strings.Contains(content, ">false<") || strings.Contains(content, ">0<") {
					return
				}
				if shouldReportMotion(cameraID) {
					fmt.Printf("[Motion] %s: motion detected (%s)\n", cameraID, topic)
					go reportMotion(cfg, cameraID)
				}
				return
			}
		}
		return
	}

	// Check if any notification message has a motion topic
	for _, msg := range env.Messages {
		topic := msg.Topic.Value
		if topic == "" {
			continue
		}

		if !isMotionTopic(topic) {
			continue
		}

		// Check for false values indicating motion stopped
		for _, item := range msg.Data.SimpleItems {
			val := strings.ToLower(item.Value)
			if val == "false" || val == "0" {
				return
			}
		}

		if shouldReportMotion(cameraID) {
			fmt.Printf("[Motion] %s: motion detected (%s)\n", cameraID, topic)
			go reportMotion(cfg, cameraID)
		}
		return
	}
}

// StartMotionDetect starts ONVIF motion detection for all configured cameras.
func StartMotionDetect(cfg *AgentConfig) {
	motionStoppedMu.Lock()
	motionStopped = false
	motionStoppedMu.Unlock()

	fmt.Println("[Motion] Motion detection listener started")
	fmt.Printf("[Motion] Monitoring %d cameras\n", len(cfg.Cameras))

	for _, camera := range cfg.Cameras {
		cam := camera // capture loop variable

		if cam.OnvifUser == "" || cam.OnvifPassword == "" {
			fmt.Printf("[Motion] %s: no ONVIF credentials, skipping ONVIF subscription\n", cam.CameraID)
			continue
		}

		hostname := extractHostFromRtsp(cam.RtspURL)
		if hostname == "" {
			fmt.Printf("[Motion] %s: could not parse hostname from RTSP URL, skipping\n", cam.CameraID)
			continue
		}

		onvifPort := 80

		fmt.Printf("[Motion] %s: connecting to ONVIF at %s:%d...\n", cam.CameraID, hostname, onvifPort)

		sub, err := onvif.Subscribe(hostname, onvifPort, cam.OnvifUser, cam.OnvifPassword, func(data []byte) {
			handleONVIFEvent(cfg, cam.CameraID, data)
		})

		if err != nil {
			fmt.Printf("[Motion] %s: ONVIF connection failed (streaming still works): %s\n", cam.CameraID, err)
			continue
		}

		activeSubscriptionsMu.Lock()
		activeSubscriptions = append(activeSubscriptions, sub)
		activeSubscriptionsMu.Unlock()

		fmt.Printf("[Motion] %s: ONVIF connected, subscribing to events\n", cam.CameraID)
	}
}

// StopMotionDetect stops all ONVIF event subscriptions and cleans up.
func StopMotionDetect() {
	motionStoppedMu.Lock()
	motionStopped = true
	motionStoppedMu.Unlock()

	fmt.Println("[Motion] Stopping motion detection...")

	activeSubscriptionsMu.Lock()
	for _, sub := range activeSubscriptions {
		sub.Close()
	}
	activeSubscriptions = nil
	activeSubscriptionsMu.Unlock()

	lastMotionTimestampsMu.Lock()
	lastMotionTimestamps = make(map[string]time.Time)
	lastMotionTimestampsMu.Unlock()

	fmt.Println("[Motion] Motion detection stopped")
}
